### ISSUE TITLE
Make InfoWindow dynamically updatable

### ISSUE THREAD

**Issue by tobrun** [mapbox/mapbox-gl-native#5237]
We expose setTitle and setSnippet on Marker to set the contents of the default InfoWindow. At this point you can only set them once. Calling these methods consecutively will not always result in changing the contents of the InfoWindow (eg. when it's already open).  We have been hearing about the requirement to do this and this follows the line set in `4.1.0` to make other Marker attributes updatable. 

We should also look into the possibility to expose an InfoWindowAdapter equivalent for this use-case by exposing an invalidate method on InfoWindowAdapter (this will result in recreating the InfoWindow).

**Comment by bleege** [comment]
I've started work on this functionality by re-enabling access to `Marker.setTitle()` and `Marker.setSnippet()` to be the API for updating this content as this is consistent with how the other dynamic marker APIs work (`Marker.setPosition()` and `Marker.setIcon()` API ).  I've also updated the test case for this in the TestApp's `DynamicMarkerChangeActivity` so that work on this can commence.  So far the InfoWindow moves from location to location just fine, but the content doesn't update.  As shown below, when the FAB is originally pressed to do the initial toggle Chelsea's content appears above Arsensal's marker and icon (sorry @jacquestardie ).

![device-2016-06-06-161305](https://cloud.githubusercontent.com/assets/989928/15838326/1389b94c-2c03-11e6-805a-10f58e8a5b23.png)
_Original State_

![device-2016-06-06-161328](https://cloud.githubusercontent.com/assets/989928/15838329/1acc1d9e-2c03-11e6-9731-e409de8cce13.png)
_State After First Toggle_

**Comment by jacquestardie** [comment]
@bleege I hope this is at the [top top top](http://www.independent.ie/sport/wenger-admits-arsenal-need-to-bring-in-top-top-top-quality-29504660.html) of your list of things to do 😄

**Comment by bleege** [comment]
I've now added the functionality to update the existing InfoWindow that uses the default InfoWindow for `setTitle()` and `setSnippet()`.  There's a bit of race condition causing the InfoWindow content to be updated a split second before the InfoWindow is moved to it's new location.  I think this has to do with the update process and the (likely unusual) use case as this case is setting all four properties at once and then calling `updateMarker()` after each property is set.

The next steps are to:

1. Verify that updated title and snippet works in InfoWindow when the InfoWindow is closed between FAB toggles
1. Verify that the updated title and snippet works in InfoWindow when the InfoWindow isn't moved
1. Try to minimize the number of `updateMarker()` calls

```java
   // Use Case Updating all four 4 items
    private void updateMarker() {
        // update model
        boolean first = (boolean) mMapView.getTag();
        mMapView.setTag(!first);

        // update marker
        mMarker.setPosition(first ? LAT_LNG_CHELSEA : LAT_LNG_ARSENAL);
        mMarker.setIcon(mIconFactory.fromResource(first ? R.drawable.ic_chelsea : R.drawable.ic_arsenal));
        mMarker.setTitle(first ? getString(R.string.dynamic_marker_chelsea_title) : getString(R.string.dynamic_marker_arsenal_title));
        mMarker.setSnippet(first ? getString(R.string.dynamic_marker_chelsea_snippet) :  getString(R.string.dynamic_marker_arsenal_snippet));
    }

// Update being run after each setXXX() method is called
MapboxMap map = getMapboxMap();
if (map != null) {
    map.updateMarker(this);
}
```

![ezgif com-video-to-gif 1](https://cloud.githubusercontent.com/assets/989928/15840513/dbc7811c-2c0f-11e6-8f78-778401d9c02d.gif)

**Comment by bleege** [comment]
> Verify that updated title and snippet works in InfoWindow when the InfoWindow is closed between FAB toggles

☑️ 

![ezgif com-video-to-gif](https://cloud.githubusercontent.com/assets/989928/15840694/2e9edcf4-2c11-11e6-96fd-7cb1d4ebadb2.gif)

**Comment by bleege** [comment]
> Verify that the updated title and snippet works in InfoWindow when the InfoWindow isn't moved

☑️ 

![location-stays-same](https://cloud.githubusercontent.com/assets/989928/15864514/13072a0a-2c9d-11e6-8deb-a07e9ac1eb20.gif)

**Comment by bleege** [comment]
Commenting out the updateMarker code from `refreshInfoWindowContent()` still updates the InfoWindow content (yay!), but I'm also still seeing a bit of lag between the two.  I'm going to try removing the updateMarker() calls that are generated by the `setPosition()` and `setIcon()` that are part of this test.

```java
            MapboxMap map = getMapboxMap();
            if (map != null) {
                map.updateMarker(this);
            }
```


![2016-06-07 11_04_02](https://cloud.githubusercontent.com/assets/989928/15865276/c53ef156-2c9f-11e6-912a-dd1e5194806d.gif)

**Comment by bleege** [comment]
Discussed the split second offset issue with @zugaldia @cammace @ivovandongen this morning and we decided to ship with this to start with and look to refine over time.  The fundamental issue is that the InfoWindow is rendered at the Java View level while the annotation is rendered in Core GL.

**Comment by bleege** [comment]
Squashed and Merged into `release-android-v4.1.0`.
